Feature: Job metrics

LuiggiTenorioK opened the pull request:
PR for #2031 

Example of usage on: https://github.com/BSC-ES/autosubmit/issues/2031#issuecomment-2545201596

### Process:

1. Should be triggered once the job finishes in COMPLETED or FAILED
2. Read the JOBS.<JOB_SECTION>.METRICS section of the job and parse it
3. Build the remote file path and read it
4. Store it on DB

LuiggiTenorioK commented on autosubmit/job/metrics_processor.py:
```diff
@@ -0,0 +1,182 @@
+
+        # For each file path, read the content of the file
+        for path, metrics_by_selector_type in metrics_by_path_selector_type.items():
+            with open(path, "r") as f:
+                content = f.read()
```
@dbeltrankyl is there any defined method to read the file's content on the remote?

I saw the `paramiko_platform.py` and there are just methods to move files from remote to local. Still, I'm confused about how to use the Platform class children correctly :face_with_spiral_eyes: 

cc @kinow

dbeltrankyl replied:
I think not. Talking by memory, I believe Autosubmit doesn't read anything from the remote; it only checks some files ( completed, failed wrapper... ) but doesn't read the content

LuiggiTenorioK commented:
Already tested it locally and seems to work as intended :+1: 

Tested by creating a local experiment with:

```yaml
JOBS:
  SIM:
    FILE: sim.sh
    METRICS:
      - NAME: metric1
        PATH: /home/ltenorio/autosubmit/a1vg/outs/
        FILENAME: a.out
      - NAME: metric2
        PATH: /home/ltenorio/autosubmit/a1vg/outs/
        FILENAME: output.json
        SELECTOR:
          TYPE: JSON
          KEY: job
```

and the `sim.sh` do:

```bash
#!/bin/bash

mkdir -p /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/

echo "%JOBNAME% $(date -Iseconds)" > /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/a.out

cat <<EOF > /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/output.json
{
    "job": "%JOBNAME%",
    "timestamp": "$(date)"
}
EOF
```

At the end, it will create a `<expid>/tmp/metrics_<expid>.db` file with a table:

| job_name  | metric_name |metric_value |
| ------------- | ------------- | ------------- |
|expid_SIM|metric1|expid_SIM 2025-02-03T14:09:26+01:00|
|expid_SIM|metric2|expid_SIM|

@dbeltrankyl how do you test new platform functions in Autosubmit? I saw that the platform module has low coverage, so I thought this part was done manually.

dbeltrankyl commented:
> @dbeltrankyl how do you test new platform functions in Autosubmit? I saw that the platform module has low coverage, so I thought this part was done manually.

Missed this comment; sorry


They are not manual, but we haven't added many platforms since we adopted this develop -> test -> feedback cycle. Most of the tests are unit tests though.

test_slurm_platform_pytest.py has some pytest
test_pjm_platform_pytest.py has some pytest

in test_paramiko.py you also have some pytest

There are some general like

test_scheduler_general.py

Usually for init a platform what I do is to write a valid config file 

Similar to 
```python
@pytest.fixture
def ps_platform(tmpdir):
    tmpdir.owner = Path(tmpdir).owner()
    config = {
        "LOCAL_ROOT_DIR": str(tmpdir),
        "LOCAL_TMP_DIR": 'tmp',
        "PLATFORMS": {
            "pytest-ps": {
                "type": "ps", # slurm, ps..
                "host": "127.0.0.1",
                "user": tmpdir.owner,
                "project": "whatever",
                "scratch_dir": f"{Path(tmpdir).name}",
                "MAX_WALLCLOCK": "48:00",
                "DISABLE_RECOVERY_THREADS": True
            }
        }
    }
    platform = PsPlatform(expid='a000', name='local-ps', config=config)
    platform.host = '127.0.0.1'
    platform.user = tmpdir.owner
    platform.root_dir = Path(tmpdir) / "remote"
    platform.root_dir.mkdir(parents=True, exist_ok=True)
    yield platform, tmpdir
    ```

dbeltrankyl reviewed (commented):
Thanks for the work @LuiggiTenorioK !

I think the code looks fine, left some questions

I have one additional question:

> 
```
#!/bin/bash

mkdir -p /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/

echo "%JOBNAME% $(date -Iseconds)" > /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/a.out

cat <<EOF > /home/ltenorio/autosubmit/a1vg/outs/%JOBNAME%/output.json
{
    "job": "%JOBNAME%",
    "timestamp": "$(date)"
}
EOF
```

Couldn't you simplify this somehow by adding a function that users call with the stuff they want and that's it? Less prone to errors

For example, for the checkpoints, I inject some stuff into the bash.cmd and users can or not call to it to the as_header function ( python, r and bash)

```python
@staticmethod
    def as_header(scheduler_header, executable):
        if not executable:
            executable = "/bin/bash"
        return textwrap.dedent("""\
            #!{0}

            """).format(executable) + \
            scheduler_header + \
            textwrap.dedent("""\
            ###################
            # Autosubmit header
            ###################
            locale_to_set=$(locale -a | grep ^C.)
            if [ -z "$locale_to_set" ] ; then
                # locale installed...
                export LC_ALL=$locale_to_set
            else
                # locale not installed...
                locale_to_set=$(locale -a | grep ^en_GB.utf8)
                if [ -z "$locale_to_set" ] ; then
                    export LC_ALL=$locale_to_set
                else
                    export LC_ALL=C
                fi 
            fi
            
            set -xuve
            job_name_ptrn='%CURRENT_LOGDIR%/%JOBNAME%'
            echo $(date +%s) > ${job_name_ptrn}_STAT_%FAIL_COUNT%

            ################### 
            # AS CHECKPOINT FUNCTION
            ###################
            # Creates a new checkpoint file upon call based on the current numbers of calls to the function
            
            AS_CHECKPOINT_CALLS=0
            function as_checkpoint {
                AS_CHECKPOINT_CALLS=$((AS_CHECKPOINT_CALLS+1))
                touch ${job_name_ptrn}_CHECKPOINT_${AS_CHECKPOINT_CALLS}
            }
            %EXTENDED_HEADER%
            ###################
            # Autosubmit job
            ###################

            """)
 ```

LuiggiTenorioK commented on autosubmit/job/metrics_processor.py:
```diff
@@ -0,0 +1,266 @@
+    # Avoid circular imports
+    from autosubmit.job.job import Job
+
+MAX_FILE_SIZE = 4 * 1024 * 1024  # Default 4MB max file size
```
Agree, I'll add that option and keep this value as the default

LuiggiTenorioK commented on autosubmit/job/metrics_processor.py:
```diff
@@ -0,0 +1,266 @@
+
+    def store_metric(self, job_name: str, metric_name: str, metric_value: Any):
+        """
+        Store the metric value in the database. Will overwrite the value if it already exists.
```
It is defined to be triggered every time the status changes to COMPLETED and FAILED. Not sure where else could it be triggered.

About having multiple values, I first wanted to have it by run but it was kinda dirty to do it, and not sure if the history module is stable enough to rely on.

dbeltrankyl replied:
> It is defined to be triggered every time the status changes to COMPLETED and FAILED. Not sure where else could it be triggered.

Hm, so the recovery command and the setstatus command could also trigger it? What would happen in that case?

> About having multiple values, I first wanted to have it by run but it was kinda dirty to do it, and not sure if the history module is stable enough to rely on.

So the user is the one responsible for permanently storing this data in another place to avoid it being overridden?

I would, maybe, at least keep track of the retrials ( with the job.fail_count, but for the vertically wrappers.. I think they're not in a failed status until all retrials all done)

I have another question, but it is more about me not entirely remembering the origin of this issue, I apologize for it; what do we want to do with these metrics? is it up to the user or the API/GUI to process/display it? Because if it is the former, I think we need to have a history of each run, no?

LuiggiTenorioK replied:
> Hm, so the recovery command and the setstatus command could also trigger it? What would happen in that case?

Yes, it should trigger it which is not bad. In any case, it will override to the same value and if the file is not there it will not write something.

> I have another question, but it is more about me not entirely remembering the origin of this issue, I apologize for it; what do we want to do with these metrics? is it up to the user or the API/GUI to process/display it? Because if it is the former, I think we need to have a history of each run, no?

Yes, the idea is to get them from the API. We don't need the history of the metric by run as an initial scope but I agree that is relevant maybe in the short term. I think if there is no rush to release the version where this feature should be included I could at it without any problem.

LuiggiTenorioK commented on autosubmit/platforms/pjmplatform.py:
```diff
@@ -29,6 +29,11 @@
 import textwrap
+
+if TYPE_CHECKING:
+    # Avoid circular imports
```
I think that's the cleanest way to do it and it actually works pretty fine :ok_hand:

LuiggiTenorioK commented on test/unit/test_user_metrics.py:
```diff
@@ -0,0 +1,292 @@
+    "metric_specs, expected",
+    [
+        (
+            {"NAME": "metric1", "PATH": "/path/to/", "FILENAME": "file1"},
```
> This indicates a remote folder, no?

Yes, it is the remote

> In line with other comments, how do we deal with job retrials?
> The second retrial will overwrite the first retrial?

Do retrials change the status to `WAITING`, right? Then it will be overwritten by the second one

> what about expid reruns?

It'll be better with versioning storing also the `run_id`. As I said, it will need to integrate with the history module to enable that.

> I re-read the issue again, and we won't have files at all locally, right? they're stored in the tmp/metrics.db

Yes, no need to copy the file, just read it. Instead, I'll have to make the process longer to arrive to the same result.

> Does the database store the timestamps?

Yes, in the modified column of the table

dbeltrankyl replied:
Thanks for the answers! 

I think that is all for my side, maybe @kinow has something else to comment

kinow replied:
Will have a look at it next week, but the only thing I have to comment right now is that there are Git conflicts :grimacing: 

![image](https://github.com/user-attachments/assets/33576d0e-1d8a-4122-a7d1-06b5213af70f)

dbeltrankyl commented on autosubmit/job/metrics_processor.py:
```diff
@@ -0,0 +1,266 @@
+                self.as_conf.normalize_parameters_keys(metric) for metric in raw_metrics
+            ]
+        except Exception:
+            Log.debug(traceback.format_exc())
```
It could work, but it seems hard to test as it affects all autosubmit code, maybe we should keep it for 4.2?

dbeltrankyl commented on autosubmit/job/metrics_processor.py:
```diff
@@ -0,0 +1,266 @@
+                Log.debug(f"Read file {path}")
+                content = content.decode(errors="replace").strip()
+            except Exception as exc:
+                Log.debug(traceback.format_exc())
```
I think it is the faster and less invasive solution.

LuiggiTenorioK commented:
Thanks @dbeltrankyl @kinow. After the first review, summarizing what are the pending tasks:

- [x] Add locale to the decode
- [x] Remove traceback
- [ ] Adding tests for the platform functions
- [ ] Allow the user to define the maximum size to read
- [ ] Get the last run_id and use it when storing the metric to have a history
- [ ] (Maybe) Move validation to the config parser

I'll ping when it is finished :grimacing:

LuiggiTenorioK commented:
@dbeltrankyl I made some changes following your comments :+1: You can review it now.

Now it allows setting a max file size to read per metric expressed in MB with 16MB as default. It is set by `MAX_READ_SIZE_MB`.

```YAML
JOBS:
  SETUP:
    METRICS:
      - NAME: model
        PATH: /remote_folder/
        FILENAME: model.txt
        MAX_READ_SIZE_MB: 8 # max 8MB to read
```

Then, the metrics are saved per experiment run and job. In the case of retrials, it will store the last one, overwriting the previous retrial.

About the validation before running, I think it can be done after not to delay the implementation of this new feature. WDYT?

Finally, I have some trouble testing it on a remote platform. So, I'll appreciate to have some help doing it :pray:

dbeltrankyl commented:
I'll take a look on Tuesday ( the github organization took me almost all day )

dbeltrankyl commented:
> I'll try to have a quick look during the next hour ( I'll be in a meeting) so expect an answer for Tuesday ( the github organization took me almost all day )

Oops, I partially reviewed this last Thursday and forgot about it due to the long weekend. I'll revise where I was  and update it today

dbeltrankyl commented:
Hello @LuiggiTenorioK , @isimo00 

I've managed to make it work after a small commit ( one if)

This is my job file:

```yaml
JOBS:
  sim:
    SCRIPT: |
      set -x
      metric_folder=%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics
      mkdir -p "$metric_folder/%JOBNAME%"
      echo "%JOBNAME% $(date -Iseconds)" > "$metric_folder/%JOBNAME%/model.txt"
      cat <<EOF > "$metric_folder/%JOBNAME%/metrics.json"
      {
          "job": "%JOBNAME%",
          "timestamp": "$(date)"
      }
      EOF
    RUNNING: once
    WALLCLOCK: 00:10
    METRICS:
#      - NAME: model
#        PATH: "%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics"
#        FILENAME: model.txt
#        MAX_READ_SIZE_MB: 8 # max 8MB to read
#      - name: metric
#        PATH: "%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics"
#        FILENAME: metrics.json
#        SELECTOR:
#          TYPE: json
#          key: job
      - NAME: model
        PATH: "/gpfs/scratch/bsc32/bsc032070/a00x/LOG_a00x/metrics"
        FILENAME: model.txt
        MAX_READ_SIZE_MB: 8 # max 8MB to read
      - name: metric
        PATH: "/gpfs/scratch/bsc32/bsc032070/a00x/LOG_a00x/metrics"
        FILENAME: metrics.json
        SELECTOR:
          TYPE: json
          key: job
```
Thanks for the work! @LuiggiTenorioK 

There is a significant issue with automating this, which is that the AS YAML List doesn't accept placeholders; we should address this for this branch...

A quicker alternative is to allow relative paths and explain it in the documentation

```yaml
JOBS:
  sim:
    SCRIPT: |
      set -x
      metric_folder=%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics
      mkdir -p "$metric_folder/%JOBNAME%"
      echo "%JOBNAME% $(date -Iseconds)" > "$metric_folder/%JOBNAME%/model.txt"
      cat <<EOF > "$metric_folder/%JOBNAME%/metrics.json"
      {
          "job": "%JOBNAME%",
          "timestamp": "$(date)"
      }
      EOF
    RUNNING: once
    WALLCLOCK: 00:10
    METRICS:
      - NAME: model
        PATH: "metrics"
        FILENAME: model.txt
        MAX_READ_SIZE_MB: 8 # max 8MB to read
      - name: metric
        PATH: "metrics" # THIS would point to "%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/"
        FILENAME: metrics.json
        SELECTOR:
          TYPE: json
          key: job
```


Results: tmp/metric_a00x.db

![image](https://github.com/user-attachments/assets/d8b9cec2-22be-4dba-906f-cbc8e41a86ab)

dbeltrankyl commented:
Another option would be to set a METRIC_FOLDER (by default, pointing to the "%HPCROOTDIR%/LOG_$EXPID" variable under DEFAULT or CONFIG and do the following changes:

1) Remove PATH from Metrics, as it is defined under Metric_folder
2) Change the script, so the user doesn't have to redefine the folder inside and just call to %METRIC_FOLDER%
3) Maybe, add a parameter["CURRENT_METRIC_FOLDER"] that points to "%METRIC_FOLDER%/%JOBNAME%"

I think this will be simpler for the user to use the metrics, and we don't need to add the feature of allowing placeholders inside lists ( for now )

What do you think?


```yaml
CONFIG:
 METRIC_FOLDER: "%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics"
JOBS:
  sim:
    SCRIPT: |
      set -x
      #metric_folder=%HPCROOTDIR%/LOG_%DEFAULT.EXPID%/metrics
      #mkdir -p "$metric_folder/%JOBNAME%"
      echo "%JOBNAME% $(date -Iseconds)" > "%METRIC_FOLDER%/%JOBNAME%/model.txt"
      cat <<EOF > "%METRIC_FOLDER%/%JOBNAME%/metrics.json"
      {
          "job": "%JOBNAME%",
          "timestamp": "$(date)"
      }
      EOF
    RUNNING: once
    WALLCLOCK: 00:10
    METRICS:
      - NAME: model
        FILENAME: model.txt
        MAX_READ_SIZE_MB: 8 # max 8MB to read
      - name: metric
        FILENAME: metrics.json
        SELECTOR:
          TYPE: json
          key: job
```

dbeltrankyl commented:
Another thing is that I don't entirely like the idea of having to define twice the same thing ( in the template and in the YAML config)

I think that the template should be able to gather the metric name from the YAML file to avoid mismatches and make it more automatic. 

So you should be able to access to the metric.type.filename from the template instead of putting "model.txt" in the template..

Maybe something like this..? 

```yaml
    METRICS:
      - NAME: model
        MAX_READ_SIZE_MB: 8 # max 8MB to read
        type: txt 
      - name: metric
        TYPE: json # not sure about this
        SELECTOR:
          #TYPE: json not sure about this
          key: job
```

1) Apply some post-processing in the autosubmit config parse inside:  `def _normalize_jobs_section(self, data_fixed: dict, must_exists: bool) -> None:)`


Which just means to add a new variable in the as_conf.experiment_data with the "name" as prefix of the parameter:

```
For metric in jobs_data[section][metrics]

name= metric[0]

as_conf.experiment_data["CURRENT_METRIC_{name}"] = "%CURRENT_ROOTDIR%/LOG_%EXPID%/%CONFIG.METRIC_FOLDER%/%JOBNAME%/{name}.{type}"
```

ex:
```
CURRENT_METRIC_model = "%CURRENT_ROOTDIR%/LOG_%EXPID%/%CONFIG.METRIC_FOLDER%/%JOBNAME%/model.txt"
CURRENT_METRIC_metric = "%CURRENT_ROOTDIR%/LOG_%EXPID%/%CONFIG.METRIC_FOLDER%/%JOBNAME%/metric).(type)"
```
*(maybe it is current_jobname, not sure?)

2) Change the script.cmd example..

```
JOBS:
  sim:
    SCRIPT: |
      set -x
      mkdir -p "%CONFIG.METRIC_FOLDER%/%JOBNAME%"
      echo "%JOBNAME% $(date -Iseconds)" > "%CURRENT_METRIC_MODEL%"
      cat <<EOF > "%CURRENT_METRIC_METRIC%"
     {
          "job": "%JOBNAME%",
          "timestamp": "$(date)"
      }
      EOF
```

LuiggiTenorioK commented:
Thanks @dbeltrankyl for the comments!

I like the idea of using an already existing placeholder when the `PATH` is defined as a relative one. I think I discussed using `%HPCROOTDIR%` with @kinow before, but I wasn't familiar with that placeholder.

> Another option would be to set a METRIC_FOLDER (by default, pointing to the "%HPCROOTDIR%/LOG_$EXPID" variable under DEFAULT or CONFIG and do the following changes:
> 
> 1. Remove PATH from Metrics, as it is defined under Metric_folder
> 2. Change the script, so the user doesn't have to redefine the folder inside and just call to %METRIC_FOLDER%
> 3. Maybe, add a parameter["CURRENT_METRIC_FOLDER"] that points to "%METRIC_FOLDER%/%JOBNAME%"
> 4. The mkdir of the folder will take place in the AS code instead of in the template

I think this option is better, but without removing the `PATH` to allow users to also give the option of defining an absolute path (maybe for global metrics that are in a shared folder in the hpc) and mixing with the `CURRENT_METRIC_FOLDER` variable. Then, 

* If `PATH` is not defined, `CURRENT_METRIC_FOLDER=%CONFIG.METRIC_FOLDER%/%JOBNAME%`
* If `PATH`  is relative, `CURRENT_METRIC_FOLDER=%CONFIG.METRIC_FOLDER%/$PATH/%JOBNAME%`
* If `PATH` is absolute, `CURRENT_METRIC_FOLDER=$PATH/%JOBNAME%`

Finally, the mkdir step seems pretty reasonable. (also, makes it easier to test)

What I'm not completely sure about is how to generate the `CURRENT_METRIC_FOLDER` variable. Do you know where we should start?

Finally, for the last comment, I'm not sure of using `CURRENT_METRIC_{name}` since the user will have to write the name or filename anyway. Then, I think having the `CURRENT_METRIC_FOLDER` variable and concatenating with the metric filename is sufficient.

dbeltrankyl commented:
> I think this option is better, but without removing the PATH to allow users to also give the option of defining an absolute path (maybe for global metrics that are in a shared folder in the hpc) and mixing with the CURRENT_METRIC_FOLDER variable. Then,


Sounds good. 

> What I'm not completely sure about is how to generate the CURRENT_METRIC_FOLDER variable. Do you know where we should start?

> Apply some post-processing in the autosubmit config parse inside: def _normalize_jobs_section(self, data_fixed: dict, must_exists: bool) -> None:)

I said here, but thinking it again,  you should do it in the update_current_parameters, because it is when you know the jobname

`Autosubmit.job.job.py
`
```
    def update_current_parameters(self, as_conf: AutosubmitConfig, parameters: dict) -> dict:
        """
        Update the %CURRENT_*% parameters with the current platform and jobs.

        :param as_conf: The Autosubmit configuration object.
        :type as_conf: AutosubmitConfig
        :param parameters: The dictionary to update with current parameters.
        :type parameters: dict
        :return: The updated parameter's dictionary.
        :rtype: dict
        """
        for key, value in as_conf.platforms_data.get(self.platform_name, {}).items():
            parameters[f"CURRENT_{key.upper()}"] = value

        for key, value in as_conf.jobs_data[self.section].items():
            parameters[f"CURRENT_{key.upper()}"] = value

        
        return parameters
```

dbeltrankyl commented:
> Apply some post-processing in the autosubmit config parse inside: def _normalize_jobs_section(self, data_fixed: dict, must_exists: bool) -> None:)

You could postprocess the metric root folder without the JOBNAME here, tho.

* If PATH is not defined, {SECTION}_METRIC_FOLDER="%CONFIG.METRIC_FOLDER%/%JOBNAME%"
* If PATH is relative, {SECTION}_METRIC_FOLDER="%CONFIG.METRIC_FOLDER%/$PATH/"
* If PATH is absolute, {SECTION}_METRIC_FOLDER="$PATH" 

>     def update_current_parameters(self, as_conf: AutosubmitConfig, parameters: dict) -> dict:

And here check the {SECTION}_METRIC_FOLDER to add the %JOBNAME% forming the %CURRENT_METRIC_FOLDER%

LuiggiTenorioK commented:
I think that only modifying the `update_current_parameters` method is enough because I'll always need the job name.

Is it necessary to also change config parser method for this? In the last case, it is hard to determine a section path since a section can have multiple paths/metrics.

dbeltrankyl commented:
> Is it necessary to also change config parser method for this? In the last case, it is hard to determine a section path since a section can have multiple paths/metrics.

No, it is not

LuiggiTenorioK commented:
> * Remove PATH from Metrics, as it is defined under Metric_folder
> * Change the script, so the user doesn't have to redefine the folder inside and just call to %METRIC_FOLDER%
> * Maybe, add a parameter["CURRENT_METRIC_FOLDER"] that points to "%METRIC_FOLDER%/%JOBNAME%"
> * The mkdir of the folder will take place in the AS code instead of in the template

Hello @dbeltrankyl, after giving some extra thought, it seems that your proposal will work pretty well because it gives more space for extension and is easier to set up a script. So, I implemented it in the last commit!

Now, the path has been removed from the metric specification, and the default base path will use `job.remote_log_dir`, which is the `LOG_{expid}` dir on some platforms. However, it will use the one defined in `CONFIG.METRIC_FOLDER`. Then, the job name will be appended to the final `CURRENT_METRIC_FOLDER`.

Now, it is easier to make the script since you can use `%CURRENT_METRIC_FOLDER%`. For example, I changed the example script like this:

```bash

METRIC_FOLDER=%CURRENT_METRIC_FOLDER%

mkdir -p $METRIC_FOLDER

echo "%JOBNAME% $(date -Iseconds)" > $METRIC_FOLDER/a.out

cat <<EOF > $METRIC_FOLDER/output.json
{
    "job": "%JOBNAME%",
    "timestamp": "$(date)"
}
EOF
```

```yaml
CONFIG:
  METRIC_FOLDER: /home/ltenorio/tmp
...
JOBS:
  SIM:
    FILE: sim.sh
    METRICS:
      - NAME: metric1
        FILENAME: a.out
      - NAME: metric2
        FILENAME: output.json
        SELECTOR:
          TYPE: JSON
          KEY: job
```

I think with that, the user experience will be better. Thanks @dbeltrankyl!

We just have to make it clear which is the real `CURRENT_METRIC_FOLDER` in the documentation, in case some user wants to make something more complicated.

dbeltrankyl commented:
Thanks @LuiggiTenorioK ! Now it looks easier to set up

dbeltrankyl commented:
Can you rebase first @LuiggiTenorioK ?

LuiggiTenorioK commented:
Rebased! Seems that there is an issue with the documentation build, but it seems to be unrelated to this PR @dbeltrankyl

LuiggiTenorioK commented:
Update: It seems that the ReadTheDocs service was having issues. Nothing from our side, it fixed itself. @dbeltrankyl

dbeltrankyl commented:
Thanks @LuiggiTenorioK 

I won't be able to look at this today as I'm investigating an issue with 4.1.14. Maybe can someone else test it? Should be similar to https://github.com/BSC-ES/autosubmit/pull/2036#issuecomment-2827435141 but changing it to reflect the new usage

dbeltrankyl commented:
Hello @LuiggiTenorioK 

I've tested it, and there are two issues.

* This part shouldn't generate the metric folder in the remote? 

```
        # Try to create the metric folder
        try:
            metric_folder.mkdir(parents=True, exist_ok=True)
        except Exception as exc:
            Log.warning(f"Failed to create metric folder {metric_folder}: {exc}")
            
```

Also, I see that the folder is being created in the script (sim.sh) that you provided in the comment. So I'm not sure what this code should do? create a folder within the root of the experiment in local, or create a folder to where the metrics are stored in the remote?


* Current_placeholders or HPC parameters in the CONFIG.METRIC doesn't seem to work.

The CURRENT placeholders are being substituted by the job.update_parameters, not the as_conf.reload()

You can check it with this conf: 

````yaml
CONFIG:
  METRIC_FOLDER: "%CURRENT_ROOTDIR%/metrics"
JOBS:
  SIM:
    SCRIPT: |
      METRIC_FOLDER=%CURRENT_METRIC_FOLDER%

      mkdir -p $METRIC_FOLDER

      echo "%JOBNAME% $(date -Iseconds)" > $METRIC_FOLDER/a.out

      cat <<EOF > $METRIC_FOLDER/output.json
      {
        "job": "%JOBNAME%",
        "timestamp": "$(date)"
      }
      EOF
    METRICS:
      - NAME: metric1
        FILENAME: a.out
      - NAME: metric2
        FILENAME: output.json
        SELECTOR:
          TYPE: JSON
          KEY: job
    running: once
    wallclock: 02:00
    
PLATFORMS:
  MARENOSTRUM5:
    TYPE: slurm
    HOST: mn5-gpp
    PROJECT: bsc32
    USER: bsc032070
    QUEUE: gp_debug
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: false
    MAX_WALLCLOCK: 48:00
    TEMP_DIR: ''
```

a) this is called by the update_status, but at this point the job.parameters seem empty

b)     def get_metric_folder(self, as_conf: AutosubmitConfig = None) -> str:

This one is called by the update status and by the update_parameters.

This function gets the config always from the as_conf, so even if the CURRENT_METRIC_FOLDER for that job in particular is updated with update_parameters and substituted with the associated platform, when it reaches that function again, it will be substituted by the original value with the placeholders.

Also, when it is called by the update_status, the job.parameters is empty

LuiggiTenorioK commented:
> Also, I see that the folder is being created in the script (sim.sh) that you provided in the comment. So I'm not sure what this code should do? create a folder within the root of the experiment in local, or create a folder to where the metrics are stored in the remote?

Sorry, I forgot to remove this snippet. My first intention was to create the folder, but I considered it better to leave the responsibility to the user in case they need preparation to obtain permissions to create those. Do you think it should be done from our side? If not, I'll remove it in my next update.

> * Current_placeholders or HPC parameters in the CONFIG.METRIC doesn't seem to work.
> 
> The CURRENT placeholders are being substituted by the job.update_parameters, not the as_conf.reload()

I didn't think of supporting CURRENT placeholders in this configuration since it looked like a circular dependency. I'll try to give a closer look at that next week.

dbeltrankyl commented:
>  Do you think it should be done from our side? If not, I'll remove it in my next update.

As you prefer, I'm okay with both things if it is documented

> I didn't think of supporting CURRENT placeholders in this configuration since it looked like a circular dependency. I'll try to give a closer look at that next week.

I think we should be able to put these variables there to be able to have versioned configuration for different platforms, otherwise we would need to hardcode the path there

LuiggiTenorioK commented:
@dbeltrankyl I've just made the changes that allow the placeholders to be substituted as expected. In the case of the template generation, it was working fine, but it was failing inside the `update_status` as you said. So, now I trigger the `update_parameters` there to get the parameters when processing the metrics.

Additionally, I removed the attempt to create the folder to be the user's responsibility.

dbeltrankyl commented:
Hello @LuiggiTenorioK 

I see

```
Error reading file %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out
Read file %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out
[WARNING] Error reading metric file at %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out: 'NoneType' object has no attribute 'decode'
Error reading file %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json
Read file %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json
[WARNING] Error reading metric file at %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json: 'NoneType' object has no attribute 'decode'
```

I also tried with` %CURRENT_ROOTDIR%` and `%CURRENT_SCRATCH_DIR%`

LuiggiTenorioK commented:
> Hello @LuiggiTenorioK
> 
> I see
> 
> ```
> Error reading file %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out
> Read file %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out
> [WARNING] Error reading metric file at %CURRENT_LOGDIR%/metrics/a01g_SIM/a.out: 'NoneType' object has no attribute 'decode'
> Error reading file %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json
> Read file %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json
> [WARNING] Error reading metric file at %CURRENT_LOGDIR%/metrics/a01g_SIM/output.json: 'NoneType' object has no attribute 'decode'
> ```
> 
> I also tried with` %CURRENT_ROOTDIR%` and `%CURRENT_SCRATCH_DIR%`

That's weird, it worked for me using `CURRENT_ROOTDIR` and `CURRENT_LOGDIR`, but running on the local platform (it didn't work with `CURRENT_SCRATCH_DIR` since it seems it is not defined in local).

Does this happen only when using the paramiko platforms?

dbeltrankyl commented:
Didn't try to use the local platform, but you can easily test it with marenostrum5 https://github.com/BSC-ES/autosubmit/pull/2036#issuecomment-2866401915 similar to this conf

LuiggiTenorioK commented:
@dbeltrankyl I tried again with MN5 and it worked. Here is my full conf: 

```yaml
DEFAULT:
  EXPID: "a1wd"
  HPCARCH: "local"
PROJECT:
  PROJECT_TYPE: local
local:
    PROJECT_PATH: /home/ltenorio/autosubmit/a1vg/random_source # I think this doesn't matter
EXPERIMENT:
  DATELIST: '20000101'
  MEMBERS: fc0
  CHUNKSIZEUNIT: month
  CHUNKSIZE: '4'
  NUMCHUNKS: '2'
  CHUNKINI: ''
  CALENDAR: standard

CONFIG:
  MAXWAITINGJOBS: 20
  TOTALJOBS: 20
  METRIC_FOLDER: "%CURRENT_ROOTDIR%/metrics"
JOBS:
  SIM:
    SCRIPT: |
      METRIC_FOLDER=%CURRENT_METRIC_FOLDER%

      mkdir -p $METRIC_FOLDER

      echo "%JOBNAME% $(date -Iseconds)" > $METRIC_FOLDER/a.out

      cat <<EOF > $METRIC_FOLDER/output.json
      {
        "job": "%JOBNAME%",
        "timestamp": "$(date)"
      }
      EOF
    METRICS:
      - NAME: metric1
        FILENAME: a.out
      - NAME: metric2
        FILENAME: output.json
        SELECTOR:
          TYPE: JSON
          KEY: job
    running: once
    wallclock: 01:00
    PLATFORM: MARENOSTRUM5
    
PLATFORMS:
  MARENOSTRUM5:
    TYPE: slurm
    HOST: mn1 # To change
    PROJECT: bsc32
    USER: bsc032595 # To change
    QUEUE: gp_debug
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: false
    MAX_WALLCLOCK: 48:00
    TEMP_DIR: ''
```

Then my template `a1wd_SIM.cmd` does have it replaced:

```bash
METRIC_FOLDER=/gpfs/scratch/bsc32/bsc032595/a1wd/metrics/a1wd_SIM
```

and the files were created as expected in GPFS.

dbeltrankyl commented:
Hello @LuiggiTenorioK 

Maybe I didn't have the latest update.. I'll revise and test it again later today. Thanks!

dbeltrankyl commented:
Hello @LuiggiTenorioK , 

It works! Not sure what happened yesterday in my main workstation, as I'm 100% sure that I did fetch and pull, but my laptop is working

![image](https://github.com/user-attachments/assets/fdd2908e-0389-45fd-9676-5922168efb78)

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -1498,6 +1503,23 @@ def update_status(self, as_conf, failed_file=False):
+                metric_procesor.process_metrics()
+            except Exception as exc:
+                # Warn if metrics are not processed
+                Log.warning(f"Error processing metrics for job {self.name}: {exc}")
```
Maybe add an error code? That explains  or points to what would be the solution

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -1546,6 +1568,32 @@ def check_completion(self, default_status=Status.FAILED, over_wallclock=False):
+            config_section: dict = as_conf.experiment_data.get("CONFIG", {})
+            base_path = Path(config_section.get("METRIC_FOLDER", base_path))
+        except Exception as exc:
+            Log.warning(f"Failed to get metric folder from config: {exc}")
```
Same

dbeltrankyl reviewed (commented):
Some very minor comments, also I think this should be explained in the RTD before merging

LuiggiTenorioK commented:
@dbeltrankyl documentation, error codes, and more tests added! I think it is ready for a final review.

dbeltrankyl reviewed (approved):
Looks good! thanks @LuiggiTenorioK

LuiggiTenorioK commented:
Thanks @dbeltrankyl! Squashing and merging...
